sweep/bucket.go

```diff
@@ -0,0 +1,186 @@
+package sweep
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnwallet"
+)
+
+const (
+	// DefaultMinFeeRate is the default minimum fee rate we'll allow when
+	// sweeping inputs. This value corresponds to a 1 sat/vbyte fee rate.
+	DefaultMinFeeRate = lnwallet.FeePerKwFloor
+
+	// DefaultMaxFeeRate is the default maximum fee rate we'll allow when
+	// sweeping inputs. This value corresponds to a 1e4 sat/vbyte fee rate.
+	DefaultMaxFeeRate = DefaultMinFeeRate + (DefaultMinFeeRate * 1e4)
+
+	// DefaultFeeRateSpacing is the default distribution of each fee rate
+	// bucket. The current value allows for a distribution as follows:
+	//
+	//   bucket #1: min=1 sat/vbyte, max=20 sat/vbyte
+	//   bucket #2: min=21 sat/vbyte, max=40 sat/vbyte...
+	DefaultFeeRateSpacing = DefaultMinFeeRate * 20
+)
+
+var (
+	// ErrBelowMinFeeRate is an error returned when a client attempts to
+	// sweep an inputs with a fee preference that results below our allowed
+	// minimum fee rate.
+	ErrBelowMinFeeRate = fmt.Errorf("fee preference below min fee rate")
+
+	// ErrExceedsMaxFeeRate is an error returned when a client attempts to
+	// sweep an inputs with a fee preference that exceeds our allowed
+	// maximum fee rate.
+	ErrExceedsMaxFeeRate = fmt.Errorf("fee preference exceeds max fee rate")
+)
+
+// bucketID denotes a bucket's identifier. This identifier is the maximum fee
+// rate allowed for any input within the bucket.
+type bucketID = lnwallet.SatPerKWeight
+
+// bucket represents a fee rate bucket which coalesces inputs with similar fee
+// rates in order to sweep them together within a single transaction to save on
+// fees.
+type bucket struct {
+	// id is the bucket's identifier, which also serves as the maximum fee
+	// rate allowed for any input within the bucket.
+	id bucketID
+
+	// inputs is the list of inputs within this bucket. These will be
+	// optimistically batched and swept together in order to save on fees.
+	inputs map[wire.OutPoint]*pendingInput
+
+	// sweepScript is the sweep output script for the sweep transaction of
+	// the inputs.
+	sweepScript []byte
+}
+
+// newBucket initializes a new bucket with the given identifier.
+func newBucket(id bucketID) *bucket {
+	return &bucket{
+		id:          id,
+		inputs:      make(map[wire.OutPoint]*pendingInput),
+		sweepScript: nil,
+	}
+}
+
+// feeRate determines the fee rate of a bucket. This is done by calculating the
+// average fee rate of all inputs within it.
+func (b *bucket) feeRate() lnwallet.SatPerKWeight {
```

Comment:
Likely can be deferred as a follow up, but we may also eventually want to introduce the concept of "strict buckets". As an example, a strict bucket of 1 sat/byte would allow a caller to express that they _only_ want that input swept at that specified fee rate and care less about batching.